lib/options_validator.js

```diff
@@ -167,4 +154,120 @@ function invalidate(validationLevel, message, logger) {
   }
 }
 
-module.exports = { assertArity, createValidationFunction };
+/**
+ * This is a class to build an operation by checking its arity and validating its options.
+ */
+class OperationBuilder {
+  constructor() {
+    this.requiredArity = -1;
+    this.validationSchema = {};
+    this.overrideOptions = {};
+  }
+
+  /**
+   * Set an operation's required arity.
+   *
+   * @method
+   * @param {number} requiredArity The operation to validate.
+   * @return {OperationBuilder} An operation builder with a set requiredArity.
+   */
+  arity(requiredArity) {
+    this.requiredArity = requiredArity;
+    return this;
+  }
+
+  /**
+   * Set an OperationBuilder's validationSchema.
+   *
+   * @method
+   * @param {object} validationSchema The schema against which to validate options.
+   * @return {OperationBuilder} An operation builder that will validate options.
+   */
+  options(validationSchema) {
+    this.validationSchema = validationSchema;
+    return this;
+  }
+
+  overrides(overrideOptions) {
+    this.overrideOptions = overrideOptions;
+    return this;
+  }
+
+  /**
+   * Build an operation by checking its arity and validating its options.
+   *
+   * @method
+   * @param {function} operationToBuild The operation to validate.
+   * @return {function} The operation with validated options.
+   */
+  build(operationToBuild) {
+    const operationBuilder = this;
+
+    function buildOperation() {
+      assertArity(arguments, operationBuilder.requiredArity);
+
+      const args = Array.prototype.slice.call(arguments);
+
+      let validationOptions = { validationLevel: VALIDATION_LEVEL_WARN };
+      if (this.s.options.validationLevel) {
```

Comment:
need to guard this more aggressively I think: `if (this.s && this.s.options && this.s.options.validationLevel) {`